### DISCUSSION TITLE
[RFC] 030 - Google Gemini 1.5 Pro API 兼容

### DISCUSSION THREAD

**Discussion by arvinxx** [lobehub/lobehub#1796] [RFCs]
## 背景

虽然在 https://github.com/lobehub/lobe-chat/pull/1776 添加了新模型支持，但是用户反馈这个方式并不能支持 1.5 模型的调用：

- #1769 
- #1794
- https://github.com/lobehub/lobe-chat/discussions/1284#discussioncomment-8942220

## 设计思路

看了下 [Google SDK 的代码](https://github.com/google/generative-ai-js/blob/c230732baa03ea19d146b598154ebbc5c87e1708/packages/main/types/requests.ts#L114-L119)，请求侧有一个 `apiVersion` 可以传参，可以设定为 `v1` 或者 `v1beta`。而且当前 1.5 的模型是只支持 用 `v1beta` 的请求调用。

```ts
/**
 * Params passed to {@link GoogleGenerativeAI.getGenerativeModel}.
 * @public
 */
export interface RequestOptions {
  /**
   * Request timeout in milliseconds.
   */
  timeout?: number;
  /**
   * Version of API endpoint to call (e.g. "v1" or "v1beta"). If not specified,
   * defaults to latest stable version.
   */
  apiVersion?: string;
}
```

所以可以用如下方式实现([refs](https://github.com/google/generative-ai-js/blob/c230732baa03ea19d146b598154ebbc5c87e1708/samples/node/advanced-function-calling.js#L1C1-L54C1))：

```ts
  // For text-only inputs, use the gemini-pro model
  const model = genAI.getGenerativeModel(
    { model: "gemini-pro" },
    { apiVersion: "v1beta" },
  );
```

那看起来可能的解决方案就是加个模型的 id 判断，如果是最新的 1.5 模型，在这设定 `apiVersion` 为 `v1beta` ：

https://github.com/lobehub/lobe-chat/blob/aaedfa716ea6d0a48266f0572d811729c7a0a25a/src/libs/agent-runtime/google/index.ts#L52-L78

然鹅我没有 api key ，所以没法测试… 看看社区有没有人可以借我个key 或者直接来实现试试的？

## 验证

如何测试 是1.5 还是 1 ？神奇咒语： 

```md
我自横刀向天笑是谁写的？
```

| Gemini 1.0 Pro | Gemini 1.5 Pro |
| --- | --- |
|![image](https://github.com/lobehub/lobe-chat/assets/28616219/915e1a87-9694-49d7-9253-18aa1125e570)| ![image](https://github.com/lobehub/lobe-chat/assets/28616219/f3798c19-c88c-419f-b0be-e202ea310ac5)|
| 1.0 稳定产出：李白，偶尔：曹操、辛弃疾、陆游 | 1.5 稳定为正确结果 |

**Comment by GalileoFe** [comment]
dc私聊你了，你看看

**Comment by arvinxx** [comment]
实际测试，个人感觉 1.5 的 api key  基本到了可以正常使用的程度了。

vision 多轮也支持，对话不 pair 也支持。

**Reply by xaya1001** [reply]
确实不错，这次更新把之前总是忽视system role的问题解决了，我还以为是gemini的问题，感谢！

**Comment by Bocchi138** [comment]
我很想知道在Google AI Studio中可使用1.5 Pro是否即代表着在其中创建的API也能访问1.5 PRO😩

**Reply by arvinxx** [reply]
不是
